lib/server/controllers/view_controller.rb

```diff
@@ -0,0 +1,128 @@
+class ViewController < Timur::Controller
+
+  # Get the tab view data. The tab view data is a json representation of a
+  # front-end layout.
+
+  def view_json
+    success(pull_view_data.to_json, 'application/json')
+  end
+
+  def update_view_json
+    unless current_user.is_admin?(@params[:project_name])
+      failure(
+        401,
+        {errors: ['You do not have the permissions to edit this data.']}
+      )
+      return
+    end
+
+    if @params.key?(:tabs)
+      @params[:tabs].each do |tab_name, tab_datum|
+        ViewTab.update(
+          @params[:project_name],
+          @params[:model_name],
+          @params[:tabs][tab_name][:name],
+          tab_datum
+        )
+      end
+    else
+      failure(
+        401,
+        {errors: ['There is no tab data to save.']}
+      )
+      return
+    end
+
+    success(pull_view_data.to_json, 'application/json')
+  end
+
+  def delete_view_json
+
+    unless current_user.is_admin?(@params[:project_name])
+      failure(
+        401,
+        {errors: ['You do not have the permissions to edit this data.']}
+      )
+      return
+    end
+
+    # Remove the view by the model name.
+    ViewTab.where({
+      project: @params[:project_name],
+      model: @params[:model_name]
+    }).destroy_all
+
+    success(pull_all_views.to_json, 'application/json')
+  end
+
+  private
+
+  def pull_view_data
+
```

Comment:
There appears to be a lot of controller logic hidden in this method and pull_all_views - as much as possible I think controller logic (e.g., things that reference @params) should be in the actual endpoint method.

Reply:
The controller logic isn't 'hidden'. The logic is in the same file and only used by that same file. I could pack  all of that logic into the endpoint function but it is my understanding that one should break down large functions into smaller parts. If the issue is how the `@params` are being used/accessed/referenced I can pass them as arguments to the sub-functions.

Please elaborate on this comment.

Reply:
pull_view_data responds to @params[:model_name] == 'all', and is called in view_json and update_view_json. However, it seems this behavior is only appropriate in the view_json context and not the update_view_json context, therefore this logic should be in the view_json controller. For clarity I'd also suggest passing in project_name and model_name as parameters. 

It also looks like both of these methods construct the same view hash. I'd consider extracting this view construction to a common method, probably on the ViewTab model. I suspect if you do this you could eliminate pull_view_data and pull_all_views entirely and fit them in one or two lines in the controller method.